### PR TITLE
Last part of addressing #415 (fwd-ported)

### PR DESCRIPTION
**Pull request by hvr**
This is a forward-port of #5180 / 7ce486e6553d24877ae3b23868b95d20bf83760a

### CODE REVIEW

File: cabal-testsuite/PackageTests/Regression/T4154/install-time-with-constraint.test.hs [hunk 1]
```diff
@@ -9,7 +9,9 @@ main = cabalTest $ withRepo "repo" $ do
 
   -- Temporarily disabled recording here because output is not stable
   recordMode DoNotRecord $ do
-      -- Constraining all uses of 'time' results in a cyclic dependency
-      -- between 'Cabal' and the new 'time'.
-      r <- fails $ cabal' "new-build" ["time", "--constraint=any.time==99999", "--dry-run"]
-      assertOutputContains "cyclic dependencies; conflict set: time:setup.Cabal, time:setup.time" r
+      r <- fails $ cabal' "new-build" ["time", "--constraint=any.time==99999", "--constraint=setup.Cabal installed", "--dry-run"]
```
Comment (grayjay) [inline cabal-testsuite/PackageTests/Regression/T4154/install-time-with-constraint.test.hs]: Why does this command need the `setup.Cabal installed` constraint now?  It seems like cabal can only choose the installed Cabal library now that it enforces a max setup Cabal version.  I thought that the two new-build commands should only differ by the constraint scope, to show that the constraint on `time` succeeds, but the constraint on `any.time` fails.

File: cabal-testsuite/PackageTests/Regression/T4154/install-time-with-constraint.test.hs [hunk 2]
```diff
@@ -9,7 +9,9 @@ main = cabalTest $ withRepo "repo" $ do
...
+      -- Constraining all uses of 'time' originally resulted in a cyclic dependency
+      -- between 'Cabal' and the new 'time':
+      -- assertOutputContains "cyclic dependencies; conflict set: time:setup.Cabal, time:setup.time" r
+      -- However, this doesn't work anymore, so instead we more directly look for:
+      assertOutputContains "time:setup.time~>time-99999 (conflict: time:setup.Cabal" r
```
Comment (grayjay) [inline cabal-testsuite/PackageTests/Regression/T4154/install-time-with-constraint.test.hs]: I'm not sure I understand this assertion.  Shouldn't it check the line that rejects the installed time due to the command line flag, now that the cycle isn't the cause of the failure?

File: cabal-testsuite/PackageTests/Regression/T4154/time.cabal [hunk 3]
```diff
@@ -4,7 +4,7 @@ cabal-version:   >=1.8
 build-type:      Custom
 
 custom-setup
-  setup-depends: base, Cabal == 99999
```
Comment (grayjay) [inline cabal-testsuite/PackageTests/Regression/T4154/time.cabal]: I think that Cabal-99999 can be deleted now that it doesn't fit the implicit setup constraints.

File: cabal-testsuite/PackageTests/Regression/T4154/install-time-with-constraint.test.hs (hunk 1 above)
Comment (hvr) [reply cabal-testsuite/PackageTests/Regression/T4154/install-time-with-constraint.test.hs]: @grayjay 

I was mostly paranoid and wanted the resulting output to be as deterministic as possible but I realize this wasn't probably the most principled way to retain the spirit of this test... 

...and I gladly take your offer on changing the tests after the PR is merged, as you clearly have a better understanding on how to write this regression test in an effective way. :-)

Comment (grayjay) [reply cabal-testsuite/PackageTests/Regression/T4154/install-time-with-constraint.test.hs]: I made a PR: #5186

**Pull request merged**